src/frame-validator.cpp

```diff
@@ -3,90 +3,106 @@
     {
-        auto vf = dynamic_cast<video_frame*>(frame);
-        if (vf == nullptr) {
+        if(_validated)
             return true;
-        }
+
+        auto vf = dynamic_cast<video_frame*>(frame);
+        if (vf == nullptr)
+            throw std::runtime_error(to_string() << "non video stream arrived to frame_validator");
+
         auto stream = vf->get_stream();
-        if (stream->get_stream_type() != RS2_STREAM_DEPTH)
+
+        //all streams except ir will be droped untill the validation on ir,
+        //after the validation all streams will be passeded to user callback directly
+        if (stream->get_stream_type() != RS2_STREAM_INFRARED)
+            return false;
+
+        //start to validate only from the second frame
+        if(frame_num++ < 2)
             return false;
 
         auto w = vf->get_width();
         auto h = vf->get_height();
-        auto bpp = vf->get_bpp() / 8;
-        auto data = reinterpret_cast<const uint16_t*>(vf->get_frame_data());
 
-        bool non_empty_pixel_found = false;
+        auto data = static_cast<const void*>(vf->get_frame_data());
+
+        auto invalid_pixels = 0;
 
-        int y_begin = h * 0.5 - h * 0.05;
-        int y_end = h * 0.5 + h * 0.05;
-        int x_begin = w * 0.5 - w * 0.05;
-        int x_end = w * 0.5 + w * 0.05;
-        for (int y = y_begin; y < y_end; y++)
+        for (int y = 0; y < h*w; y++)
         {
-            for (int x = x_begin; x < x_end; x++)
+            if (((byte*)data)[y] == 0)
             {
-                auto index = x + y * w;
-                if (data[index] != 0)
-                {
-                    non_empty_pixel_found = true;
-                    break;
-                }
+                invalid_pixels++;
             }
         }
 
-        if (non_empty_pixel_found)
+        if ((float)invalid_pixels/(w*h) < INVALID_PIXELS_THRESHOLD)
+        {
+            _validated = true;
             return true;
+        }
         else
         {
-            LOG_ERROR("frame_source received an empty depth frame, restarting the sensor...");
+
+            LOG_ERROR("frame_source received corrupted depth frame ("<<(float)invalid_pixels/(w*h)<<"% invalid pixels), restarting the sensor...");
             auto s = _sensor;
-            auto cr = _current_requests;
+            auto vr = _validator_requests;
             auto uc = _user_callback;
-            _reset_thread = std::thread([s, cr, uc]()
+            _stopped = true;
+            _reset_thread = std::thread([s, vr, uc]()
             {
                 try {
+                    std::this_thread::sleep_for(std::chrono::milliseconds(500));
                     s->stop();
                     s->close();
-                    s->open(cr);
+                    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+                    s->open(vr);
                     s->start(uc);
                 }
-                catch (...) {}
+                catch (...)
```

Comment:
If an exception is raised during the `stop->close->open->start` then the system state is not defined. 
the thread's shall complete with either streaming or a full stop. In both cases a log shall be provided